### PR TITLE
[WSL-359] Implemented distro.SetAsDefault and DefaultDistro

### PR DESCRIPTION
**Pull request by EduardGomezEscandell**
A getter and a setter for the default distro

### CODE REVIEW

File: wslexe_windows.go
Comment (EduardGomezEscandell) [inline wslexe_windows.go:37]: TODO: Use registry rather than starting a whole process to set this field.

File: wslexe_windows.go
Comment (EduardGomezEscandell) [inline wslexe_windows.go:45]: This is hard to exercise without breaking the testbed (since it would require unregistering all distros, which I'm not willing to do on my machine).

File: wslexe_windows.go
Comment (EduardGomezEscandell) [reply wslexe_windows.go:37]: Cannot do: editing the registry requires special permissions. I implemented this function via the registry and got:
```
failed to set "wsltesting_TestDistroSetAsDefault_93405798" as default distro: Access is denied.
```

Comment (didrocks) [reply wslexe_windows.go:37]: Maybe add it as a comment in the function for documentation and not wondering again why we can’t.

There are some policy-kit model I think to get special permissions, as `wsl.exe --set-default` can be run without being admin? (not really suitable for a package though as it means every process would need to register with those special perms)

File: distro_test.go
Comment (didrocks) [inline distro_test.go:77]: can you add a comment on what we nolint?

File: distro_test.go
Comment (didrocks) [inline distro_test.go:78]: This is some kind of "Setup:", no?

File: distro_test.go
Comment (didrocks) [inline distro_test.go:93]: I would be more descriptive in the test name.

"happy path" -> "succesfully set existing distro as default"
"fake distro" -> "error on setting unexisting distro as default"

File: distro_test.go
Comment (didrocks) [inline distro_test.go:106]: Nipick: I would add a line break just above this to separate command/main error/no error assertions, and then, those additional "check default" assertions.

File: main_test.go
Comment (didrocks) [inline main_test.go:34]: perfect on those setup() returning the teardown() :)

File: main_test.go
Comment (didrocks) [inline main_test.go:217]: niptick: I would add some line breaks in that function by group of functionality :)
```
Command + immediate error check (string contains and err != nil)

Parsing output with scanner

return
```

File: wslexe_windows.go
Comment (EduardGomezEscandell) [reply wslexe_windows.go:37]: Oh no, `wsl.exe --set-default` works fine. What doesn't work is the TODO.

AFAIK all  `wsl.exe --set-default` does is change the field Lxss\DefaultDistribution in the Windows registry. So I tried to avoid spawning the wsl.exe process by modifying the registry directly. This is not allowed. So I fell back to the solution you see here.

File: main_test.go
Comment (didrocks) [inline main_test.go:211]: shouldn’t we check `err != nil` before getting `out`? We can’t ensure the first parameter has valid content if it errored out.

File: wslexe_windows.go
Comment (EduardGomezEscandell) [reply wslexe_windows.go:37]: Yeah I think that is possible, the issue is the one you present. I think sticking to this sub-optimal solution is fine.

File: main_test.go
Comment (didrocks) [inline main_test.go:249]: ditto on gosec nolint.

File: wslexe_windows.go
Comment (didrocks) [reply wslexe_windows.go:45]: understood :)

File: wslexe_windows.go
Comment (didrocks) [inline wslexe_windows.go:47]: If you feel you are repeating this too much, feel free to copy the decorate function from adsys (it’s a small copy which is better than a huge dependency)

**Review by didrocks — CHANGES_REQUESTED** [review]
Some small comments to enhance the readability and maintainability of the code in general. Otherwise, very well done (as usual? :p). In particular on looking at the registry ordering (hoping it’s stable though!)

File: wslexe_windows.go
Comment (didrocks) [reply wslexe_windows.go:37]: agreed :)

File: distro_test.go
Comment (EduardGomezEscandell) [reply distro_test.go:77]: All the exec.Command //nolint are `G204: Subprocess launched with a potential tainted input or cmd arguments`.

I'll add the comment.

File: distro_test.go
Comment (EduardGomezEscandell) [reply distro_test.go:78]: Gotcha 😄

File: main_test.go
Comment (EduardGomezEscandell) [reply main_test.go:211]: Well that message will only be printed when an error is thrown. I will make it safer by ensuring the error is `exec.ExitError` before reading the output.

File: distro_test.go
Comment (didrocks) [reply distro_test.go:77]: you can add a rationale like "we control the argument in tests" or something like that.

File: wslexe_windows.go
Comment (EduardGomezEscandell) [reply wslexe_windows.go:47]: I added it to my gardening ideas :D
https://github.com/users/EduardGomezEscandell/projects/1/views/1

**Comment by EduardGomezEscandell** [comment]
Force-pushed to rebase main

File: main_test.go
Comment (didrocks) [reply main_test.go:258]: I think you can prefix the paragraph then with the nolint gosec? Something like this: https://github.com/ubuntu/adsys/blob/main/internal/policies/privilege/privilege.go#L92.

This way. the comment and the notlint directive are in the same paragraph and are connected.

File: distro_test.go
Comment (didrocks) [inline distro_test.go:97]: I see you like to say fail, but it does return an error, this is really a nitpick of mine feel free to ignore :)

File: wslexe_windows.go
Comment (didrocks) [reply wslexe_windows.go:55]: \o/

**Review by didrocks — APPROVED** [review]
Still some suggestions on the recent changes (one really nitpick that you can ignore, and another one to make it more readable and maintainable IMHO as we both remove the lint and comment or none).

Once you comment/fix, feel free to merge :)

File: main_test.go
Comment (EduardGomezEscandell) [reply main_test.go:258]: Oh this looks much nicer. Done 👍.

File: distro_test.go
Comment (EduardGomezEscandell) [reply distro_test.go:97]: We understand failure as an unreported error, then?

Comment (didrocks) [reply distro_test.go]: To me, yes, but it can be a french thingy :p

**Comment by EduardGomezEscandell** [comment]
Rests passed locally.
I can't wait for automated tests (https://github.com/EduardGomezEscandell/GoWSL/pull/10) 😅

**Comment by didrocks** [comment]
> I can't wait for automated tests (https://github.com/EduardGomezEscandell/GoWSL/pull/10) sweat_smile
The one you are not brave enough to submit? :p

**Comment by EduardGomezEscandell** [comment]
> The one you are not brave enough to submit? :p

Haha I'm waiting for its dependency in Ubuntu/WSl to be merged

File: distro_test.go
Comment (EduardGomezEscandell) [reply distro_test.go]: Haha well it makes sense so let's stick to this nomenclature.

**Pull request merged**
